### PR TITLE
[feat] Add ProfileEvents map to PartLog

### LINKED ISSUE #10316: metrics in system.part_log
Add columns ProfileEvents.Names, ProfileEvents.Values for system.part_log.
This will help to track resource usage for merges.

This task is needed for Yandex Morda.

### PR DESCRIPTION
**Pull request by johnskopis**
closes #10316

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Expose ProfileEvents counters in system.part_log

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by johnskopis** [comment]
hey @alexey-milovidov do you mind adding some guidance here?

Is it OK to use attach/detach like this?

This commit is a bit off the mark...but what I was thinking:

Task should be created with a TaskObserver.
A TaskObserver is constructed with different underlying observers:
```
TaskObserver.make(TaskObserver::ProfileEvents, TaskObserver::OpenTracing, ...)
// or
TaskObserver().add(TaskObserverProfileEvents()).add(...)
```

This allows for the config to specify which observers to enable, and what sample rate (if any) to use.


When the task is switched out/in call observer.onSuspend/observer.onResume, which in turn calls the onSuspend/onResume of the underlying observer:
```
TaskObserver::onSuspend()
{
 for (auto observer: observers) 
  observer.onSuspend()
}
```

Currently the data (ThreadGroupStatusPtr) used with this concrete observer is stored on itself. Can/should we move this to TaskRuntimeData? Or does it make more sense to store a map <STEP,ProfileEvents::Counters::Snapshot> on the TaskRuntimeData?

What do you think about updating PartLogWriter to take a TaskRuntimeDataPtr? It seems like PartLogWriter should be called ({}, runtime_data) from onComplete. Before calling the Task.onCompleted call TaskObserver.onCompleted to finalize any data in the shared TaskRuntimeDataPtr

Also, it doesn't always make sense to include ProfileEvents on part_log events. For example, a NewPart event includes the ProfileEvents already in the query log so I'm not even sure it's the right approach. 

Thanks for all you do :)

**Comment by novikd** [comment]
@Mergifyio update

File: src/Storages/MergeTree/TaskObserverMetrics.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+class TaskObserverMetrics : public boost::noncopyable
+{
+public:
+  TaskObserverMetrics() : thread_group(std::make_shared<ThreadGroupStatus>()) {}
+  ~TaskObserverMetrics() {}
+
+  bool doResume()
+  {
+      CurrentThread::attachTo(thread_group);
```
Comment (novikd) [inline src/Storages/MergeTree/TaskObserverMetrics.h:21]: Why do you need attach?

**Comment by johnskopis** [comment]
I'm not able to sign the CLA. Sorry. Thanks bye

**Comment by alexey-milovidov** [comment]
@johnskopis Instead of signing a CLA you can copy-paste this text verbatim to a comment:

```
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

**Comment by johnskopis** [comment]
(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file

Since creating this PR my situation has changed (I started a new job on Feb 28) so I will need to abstain.

Hope it makes sense

**Comment by nvartolomei** [comment]
Resubmitted as #35989

**Pull request closed without merging**